File: poc/boost.python/diamond/diamond.py
import poc
from _diamond import A as _A, B as _B

class A(_A):
    # Call the C++ inherited constructor and set cxxinit.
    def __init__(self, _i):
        print('Python A constructor start')
        if not hasattr(self, 'cxxinit'):
            _A.__init__(self)
            self.cxxinit = True
        print('Python A constructor end')

    # Only the Python class where a certain C++ method is provided for
    # the first time calls the C++ method.
    def doIt(self, i):
        print('Python A doIt start')
        _A.doIt(self)
        print('Python A doIt end')

class B(A, _B):
    # First call the C++ inherited constructor and set cxxinit, only
    # then call the Python inherited constructor.
    def __init__(self):
        print('Python B constructor start')
        if not hasattr(self, 'cxxinit'):
            _B.__init__(self)
            self.cxxinit = True
        A.__init__(self, 52)
        print('Python B constructor end')

    # Do NOT call the C++ method, but the Python inherited method.
    def doIt(self):
        print('Python B doIt start')
        A.doIt(self, 42)
        print('Python B doIt end')

b = B()
b.doIt()
b.doIt2()

File: poc/boost.python/diamond/diamond.cpp
#include <boost/python.hpp>
#include <iostream>

using namespace std;
using namespace boost::python;

class A 
{
public:
  A() {
    cout << "C++ A constructor" << endl;
  }
  
  virtual void doIt2() {
    cout << "C++ A doIt2" << endl;
  }

  virtual void doIt() {
    cout << "C++ A doIt start" << endl;
    cout << "C++ A doIt end" << endl;
  }

  virtual ~A() {
    cout << "C++ A destructor" << endl;
  }
};

class B 
  : public A 
{
public:
  B() {
    cout << "C++ B constructor" << endl;
  }

  virtual void doIt() {
    cout << "C++ B doIt start" << endl;
    this->A::doIt();
    cout << "C++ B doIt end" << endl;
  }
};

BOOST_PYTHON_MODULE(_diamond)
{
  class_< A, boost::noncopyable >("A")
    .def("doIt", &A::doIt)
    .def("doIt2", &A::doIt2)
    ;

  class_< B, bases< A >, boost::noncopyable >("B")
    .def("doIt", &B::doIt)
    ;
}
